snapshot/git/gitdb/bundlestore.go

```diff
@@ -0,0 +1,200 @@
+	if len(extraParts) != 3 {
+		return nil, fmt.Errorf("cannot parse snapshot ID: expected 5 parts in bundlestore id: %s", id)
+	}
+	bundleName, streamName, sha := extraParts[0], extraParts[1], extraParts[2]
+
+	if err := validSha(sha); err != nil {
+		return nil, err
+	}
+
+	return &bundlestoreSnapshot{kind: kind, sha: sha, bundleName: bundleName, streamName: streamName}, nil
+}
+
+func (b *bundlestoreBackend) upload(s snapshot, db *DB) (snapshot, error) {
+	// We only have to upload a localSnapshot
+	switch s := s.(type) {
+	case *tagsSnapshot:
+		return s, nil
+	case *streamSnapshot:
+		return s, nil
+	case *bundlestoreSnapshot:
+		return s, nil
+	case *localSnapshot:
+		return b.uploadLocalSnapshot(s, db)
+	default:
+		return nil, fmt.Errorf("cannot upload %v: unknown type %T", s, s)
+	}
+}
+
+// git bundle create takes a rev list; it requires that it include a postiive ref
+// so we can't just do:
+// git bundle create 545c88d71d40a49ebdfb1d268c724110793330d2..3060a3a519888957e13df75ffd09ea50f97dd03b
+// instead, we have to write a temporary ref
+const bundlestoreTempRef = "reserved_scoot/bundlestore/__temp_for_writing"
+
+func (b *bundlestoreBackend) uploadLocalSnapshot(s *localSnapshot, db *DB) (sn snapshot, err error) {
+	// the sha of the commit we're going to use as the ref
+	commitSha := s.sha
+
+	// the revList to create the bundle
+	// unless we can find a better merge base, we'll just include the commit
+	revList := bundlestoreTempRef
+
+	// the name of the stream that this bundle requires
+	streamName := ""
+
+	switch s.kind {
+	case kindGitCommitSnapshot:
+		if db.stream.cfg != nil && db.stream.cfg.RefSpec != "" {
+			streamHead, err := db.dataRepo.Run("rev-parse", db.stream.cfg.RefSpec)
+			if err != nil {
+				return nil, err
+			}
+
+			mergeBase, err := db.dataRepo.RunSha("merge-base", streamHead, commitSha)
+
+			// if err != nil, it just means we don't have a merge-base
+			if err == nil {
+				revList = fmt.Sprintf("%s..%s", mergeBase, bundlestoreTempRef)
+				streamName = db.stream.cfg.Name
+			}
+
+		}
+	case kindFSSnapshot:
+		commitSha, err = db.dataRepo.RunSha("commit-tree", commitSha, "-m",
+			"commit to distribute GitDB FSSnapshot via bundlestore")
+	default:
+		return nil, fmt.Errorf("unknown Snapshot kind: %v", s.kind)
+	}
+
+	// update the ref
+	if _, err := db.dataRepo.Run("update-ref", bundlestoreTempRef, commitSha); err != nil {
+		return nil, err
+	}
+
+	d, err := db.tmp.TempDir("bundle-")
+	if err != nil {
+		return nil, err
+	}
+
+	// we can't use tmpDir.TempFile() because we need the file to not exist
+	bundleFilename := path.Join(d.Dir, fmt.Sprintf("commit-%s.bundle", s.SHA()))
+
+	// create the bundle
+	if _, err := db.dataRepo.Run("bundle", "create", bundleFilename, revList); err != nil {
+		return nil, err
+	}
+
+	f, err := os.Open(bundleFilename)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	if err := b.cfg.Store.Write(s.sha, f); err != nil {
```

Comment:
This seems to be one of the few places that refers to snapshot data without %s-%s-%s. It's one thing to checkout a raw sha1, but I'm not sure it makes sense to do the same elsewhere.